embeddings/meta2vec.py

```diff
@@ -0,0 +1,294 @@
+        for example, occur when the first or last node context is extracted.
+        :param padding_value: the value it should be set to.
+        :return: the object itself.
+        """
+        self.padding_value = padding_value
+        return self
+
+    @classmethod
+    def from_json(cls, json, seperator=" | ") -> 'MetaPathsInput':
+        converted_meta_paths = []
+        node_types = set()
+        for meta_paths in json.keys():
+            node_ids = [int(id) for id in meta_paths.split(seperator)]
+            converted_meta_paths.append(node_ids)
+            node_types |= set(node_ids)
+
+        return cls(converted_meta_paths, node_types)
+
+    def _apply_transformation(self, meta_paths):
+        features = {'node': [], 'context': []}
+        for paths in meta_paths:
+            for node_key in range(len(paths)):
+                node = paths[node_key]
+                left_keys = self._left_context(node_key, self.window_size)
+                right_keys = self._right_context(node_key + 1, len(paths), self.window_size)
+                context_keys = np.array(left_keys + right_keys) + 1
+                context = np.array([self.padding_value - 1] + paths, dtype=np.float32) + 1
+                context = context[context_keys]
+
+                features['node'].append(node)
+                features['context'].append(context)
+        # Finally convert to array
+        features['context'] = np.array(features['context'])
+        return features['node'], features['context']
+
+    @staticmethod
+    def _left_context(max_key, win_size):
+        return MetaPathsInput._primitive_context(list(range(max_key)), win_size)
+
+    @staticmethod
+    def _right_context(min_key, max_key, win_size):
+        return MetaPathsInput._primitive_context(list(range(min_key, max_key)), win_size)
+
+    @staticmethod
+    def _primitive_context(key_range, windows_size):
+        if len(key_range) < windows_size:
+            return [-1] * (windows_size - len(key_range)) + key_range
+        return random.sample(key_range, windows_size)
+
+    def _update(self):
+        if len(self.nodes) < 1:
+            self.nodes, self.contexts = self._apply_transformation(self.meta_paths)
+
+    def skip_gram_input(self) -> tf.data.Dataset:
+        """
+        Get the dataset to train on in skip-gram format.
+        :return: the dataset with nodes as features and context as labels.
+        """
+        self._update()
+        return tf.data.Dataset().from_tensor_slices(({'node': self.nodes}, self.contexts))
+
+    def bag_of_words_input(self) -> tf.data.Dataset:
+        """
+        Get the dataset to train on in continuous bag of words format.
+        :return: the dataset with context as features and nodes as labels.
+        """
+        self._update()
+        return tf.data.Dataset().from_tensor_slices(({'context': self.contexts}, self.nodes))
```

Comment:
Shouldn't this return the node types of the nodes? Or better introduce a parameter for node or node type.